src/runtime/metal.cpp

```diff
@@ -714,18 +734,42 @@ WEAK int halide_metal_run(void *user_context,
     halide_assert(user_context, state_ptr);
     module_state *state = (module_state*)state_ptr;
 
-    mtl_function *function = new_function_with_name(state->library, entry_name, strlen(entry_name));
-    if (function == 0) {
-        error(user_context) << "Metal: Could not get function " << entry_name << "from Metal library.\n";
-        return -1;
+    
+    cached_pipeline_state *cached_pipeline = state->cached_pipeline_state_list;
+    while (cached_pipeline != NULL && strcmp(cached_pipeline->function_name, entry_name) != 0) {
+        cached_pipeline = cached_pipeline->next;
     }
 
-    mtl_compute_pipeline_state *pipeline_state = new_compute_pipeline_state_with_function(metal_context.device, function);
-    if (pipeline_state == 0) {
-        error(user_context) << "Metal: Could not allocate pipeline state.\n";
-        return -1;
+    
+    if(!cached_pipeline){
```

Comment:
Why not just create the pipeline states during `halide_metal_initialize_kernel()`?

Reply:
Also, spacing.